File: C&C++題目/Problem19_surrounded regions/Problem19_surrounded regions.cpp
#include <iostream>
#include <vector>
#include <string>
#include <set>

using namespace std;


class Solution {
public:
    void solve(vector<vector<char>>& board) {
        for(int i=0; i<board.size(); i++){
            for(int j=0; j<board[i].size(); j++){
                if(board[i][j] == 'O'){
                    vector<vector<int>> loc;
                    set<vector<int>> compare;
                    loc.push_back({i,j});
                    compare.insert({i,j});
                    int ptr = 0;
                    int flip = 1;
                    while(ptr < loc.size()){
                        int r = loc[ptr][0];
                        int c = loc[ptr][1];
                        vector<vector<int>> aside = {{r+1, c}, {r-1,c}, {r, c+1}, {r, c-1}};
                        if(r-1 < 0 || c-1 <0 || r+1 >= board.size() || c+1>=board[i].size()){
                            flip = 0;
                            break;
                        }
                        for(int v=0; v<aside.size(); v++){
                            int row = aside[v][0];
                            int col = aside[v][1];
                            if(board[row][col] == 'O'){
                                int pre = compare.size();
                                compare.insert({row, col});
                                if(pre+1 == compare.size()){
                                    loc.push_back({row, col});
                                }                              
                            }
                        }
                        ptr++;
                    }

                    if(flip){
                        for(int v=0; v<loc.size(); v++){
                            int row = loc[v][0];
                            int col = loc[v][1];
                            board[row][col] = 'X';
                        }
                    }

                    
                }
                       
            }
        }

    }
};

int main()
{

    vector<vector<char>> board = {{'X', 'X', 'X', 'X'}, {'X', 'O', 'O', 'X'}, {'X', 'X', 'O', 'X'}, {'X', 'O', 'X', 'X'}};
    Solution solution;
    solution.solve(board);
    return 0;
}

File: TypeScript題目/Problem19_surrounded regions/Problem19_surrounded regions.ts

function solve(board: string[][]): void {
    //用遞迴的方式進行DFS更快
    function DFS(row:number, col:number): void{
        if(row < 0 || row == height || col < 0 || col == width){
            return;
        }
        if(board[row][col] == "O"){
            board[row][col] = "N";
            DFS(row+1,col);
            DFS(row-1,col);
            DFS(row,col+1);
            DFS(row,col-1);
          
        }
    };
    //先遍歷外圍，出現"O"時利用DFS將其與其相連的"O"都改成"N"表示不會被翻轉
    var height = board.length;
    var width = board[0].length;
    //上圍&下圍
    for(let col=0; col<width; col++){
        if(board[0][col] == "O"){
            DFS(0, col);
        }
        if(board[height-1][col] == "O"){
            DFS(height-1, col);
        }
    }
    //左圍&右圍
    for(let row=0; row<height; row++){
        if(board[row][0] == "O"){
            DFS(row, 0);
        }
        if(board[row][width-1] == "O"){
            DFS(row, width-1);
        }
    }
    
    //遍歷borad，將"O"翻轉成"X"並將"N"轉回"O"
    for(let row=0; row<height; row++){
        for(let col=0; col<width; col++){
            if(board[row][col] == "O"){
                board[row][col] = "X";
            }
            else if(board[row][col] == "N"){
                board[row][col] = "O";
            }
        }
    }
    
};


var board: string[][] = [["X","X","X","X"],["X","O","O","X"],["X","X","O","X"],["X","O","X","X"]];
solve(board);
console.log(board)

